shared/src/tx_pool_ext.rs

```diff
@@ -0,0 +1,394 @@
+use crate::snapshot::Snapshot;
+use crate::tx_pool::{DefectEntry, PoolError, ProposedEntry, TxPool};
+use ckb_core::block::Block;
+use ckb_core::cell::{
+    resolve_transaction, OverlayCellProvider, ResolvedTransaction, UnresolvableError,
+};
+use ckb_core::transaction::{ProposalShortId, Transaction};
+use ckb_core::Cycle;
+use ckb_dao::DaoCalculator;
+use ckb_logger::{debug_target, error_target, trace_target};
+use ckb_store::ChainStore;
+use ckb_util::LinkedFnvHashSet;
+use ckb_verification::{ContextualTransactionVerifier, TransactionVerifier};
+use lru_cache::LruCache;
+use numext_fixed_hash::H256;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+impl TxPool {
+    pub fn add_txs_to_pool(&mut self, txs: Vec<Transaction>) -> Result<Vec<Cycle>, PoolError> {
+        txs.into_iter()
+            .map(|tx| self.add_tx_to_pool(tx, None))
+            .collect()
+    }
+
+    // Add a verified tx into pool
+    // this method will handle fork related verifications to make sure we are safe during a fork
+    pub fn add_tx_to_pool(
+        &mut self,
+        tx: Transaction,
+        cycles: Option<Cycle>,
+    ) -> Result<Cycle, PoolError> {
+        let tx_size = tx.serialized_size();
+        if self.reach_size_limit(tx_size) {
+            return Err(PoolError::LimitReached);
+        }
+        let short_id = tx.proposal_short_id();
+        match self.resolve_tx_from_pending_and_proposed(&tx) {
+            Ok(rtx) => self.verify_rtx(&rtx, cycles).and_then(|cycles| {
+                if self.reach_cycles_limit(cycles) {
+                    return Err(PoolError::LimitReached);
+                }
+                if self.contains_proposed(&short_id) {
+                    if let Err(e) = self.proposed_tx_and_descendants(Some(cycles), tx_size, tx) {
+                        debug_target!(
+                            crate::LOG_TARGET_TX_POOL,
+                            "Failed to add proposed tx {:?}, reason: {:?}",
+                            short_id,
+                            e
+                        );
+                        return Err(e);
+                    }
+                    self.update_statics_for_add_tx(tx_size, cycles);
+                } else if self.enqueue_tx(Some(cycles), tx_size, tx) {
+                    self.update_statics_for_add_tx(tx_size, cycles);
+                }
+                Ok(cycles)
+            }),
+            Err(err) => Err(PoolError::UnresolvableTransaction(err)),
+        }
+    }
+
+    fn contains_proposed(&self, short_id: &ProposalShortId) -> bool {
```

Comment:
The function name is confusing, maybe we can remove this function, only line 147 call this function.